### PR TITLE
Feature/445 read only offline mode for plant layer

### PR DESCRIPTION
**Pull request by Bushuo**
## Basics

- [ ] I added a line to [/doc/CHANGELOG.md](/doc/CHANGELOG.md)
- [ ] The PR is rebased with current master.
- [ ] Details of what you changed are in commit messages.
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.

## Checklist

- [ ] I have installed and I am using [pre-commit hooks](../doc/contrib/README.md#Hooks)
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added unit tests for my code
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](https://github.com/ElektraInitiative/PermaplanT/blob/master/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

### CODE REVIEW

**Comment by Bushuo** [comment]
@markus2330 
I encountered a problem with the offline functionality.
If the access token expires during offline, we need to sign in again.
The automatic refresh token mechanism can not be triggered manually and does not work.

Overall it works but logs a lot of error messages and toasts.
The keycloak client we use on the frontend does not have the control I would like it to have.

**Comment by markus2330** [comment]
> I encountered a problem with the offline functionality.
> If the access token expires during offline, we need to sign in again.
> The automatic refresh token mechanism can not be triggered manually and does not work.

We later (probably after thesis) could look into the `Offline Session` feature of KeyCloak. Of course it only works if users go offline on purpose.

> Overall triggering the singin again works, but logs a lot of error messages and toasts.
> The keycloak client we use on the frontend does not have the control I would like it to have.

There is also another one, we had only a quite short discussion about it and unfortunately we didn't write a decision. Hopefully we can talk with @absurd-turtle about that soon.

In general you don't need to implement a solution for all the problems you encounter (e.g. also SSE). In the thesis you could write about these problems and potential solutions (implementation challenges).

**Review by markus2330 — APPROVED** [review]
Works for me, when toggling Firefox to be offline, read-only mode gets activated.

If already in planting mode, you can still plant plants though. In general it seems like planting does not check errors from (not reachable) backend?

PR needs rebase though.

**Comment by Bushuo** [comment]
@markus2330 
Yes they are not handled yet, and the UI state is updated immediately without waiting for the server's response.
What i tried now and what works (apart from a little UI flashing) is caching a previous snapshot of the state that gets applied if the request fails.

I hope that this approach will not cause race conditions. A more elaborate solution that works with the same responsiveness is pretty complex.

```ts
function executeAction(action: Action<unknown, unknown>, set: SetFn, get: GetFn) {
  // get the current state before executing the action
  const snapshot = get();
  
  trackUserAction(action, set);

  action.execute(get().untrackedState.mapId).catch(() => {
    // apply the snapshot if the execution of the actions fails.
    set(() => ({
      ...snapshot,
    }));
  });
  
  // code further down is called synchronous.
  // .catch is called asynchronously
  
  trackReverseActionInHistory(action, get().step, set, get);
  applyActionToStore(action, set, get);

  set((state) => ({
    ...state,
    step: state.step + 1,
    canRedo: false,
    canUndo: true,
    history: state.history.slice(0, state.step + 1),
  }));
}
```

**Comment by Bushuo** [comment]
@markus2330
> From what I understand is the "offline-as-told-by-the-browser" and "offline-as-told-by-failed-API-requests" totally different things?

Yes they are.
The browser/network connectivity is observed by a native browser api.
The backend being unavailable can not be observed.

> Sounds great, flickering is no deal, as this usually should not happen anyway. I am not sure if we should put this also within this PR.

I already included it in the latest commit.
I will add read-only in planting mode to this PR and update the status then.

**Review by ThorbenStaufer — APPROVED** [review]
Looks good to me. I really like, that the unavailable buttons are greyed out.

**Review by 4ydan — COMMENTED** [review]
Besides many error alerts, I couldnt see any of the changes. I tried it with an offline extension or with shutting down the backend.

**Comment by markus2330** [comment]
@4ydan with Firefox, Press Alt -> File -> "Work offline" worked.

Really going offline (disconnecting computer) could be tested, too.

File: frontend/src/features/map_planning/layers/plant/PlantsLayer.tsx [hunk 1]
```diff
@@ -15,20 +16,27 @@ import { Layer } from 'react-konva';
 import { Html } from 'react-konva-utils';
 import * as uuid from 'uuid';
 
+const PLANT_WIDTHS = new Map([
+  ['narrow', 10],
+  ['medium', 50],
+  ['wide', 100],
+]);
```
[the remark this answers is not in the document]
Comment (markus2330) [reply frontend/src/features/map_planning/layers/plant/PlantsLayer.tsx:23]: Thx, good suggestion for the current code but just for you to know: we want to get rid of PlantSpread, see #822

Btw. in GitHub you can have several comments in one review.

File: frontend/src/features/map_planning/store/TrackedMapStore.ts [hunk 2]
```diff
@@ -228,3 +233,35 @@ function redo(set: SetFn, get: GetFn): void {
...
+    if (typeof value === 'object' && isRef(value)) {
+      continue;
+    }
+
+    clone[key] = JSON.parse(JSON.stringify(value));
+  }
+
+  return clone;
+}
+
```
[the remark this answers is not in the document]
Comment (Bushuo) [reply frontend/src/features/map_planning/store/TrackedMapStore.ts:267]: We need it for reverting if an action fails due to the network.
For those cases we could implement a serializer.
We can view it as technical debt, until we find a better solution.

[the remark this answers is not in the document]
Comment (Bushuo) [reply frontend/src/features/map_planning/store/TrackedMapStore.ts]: Actually using the undo function there would also work^^

**Review by markus2330 — APPROVED** [review]
(Looked at it together)

**Comment by Bushuo** [comment]
@markus2330 I added the changes that I wanted.
I think this can be merged now and I will move it to 'In Review'

**Comment by markus2330** [comment]
Thank you so much! :heart: Amazing work.

**Pull request merged**